Add bindings for 'zmq_poll'.

civodul opened the pull request:
Hi again!

This adds bindings for `zmq_poll`.  This allows you to wait for read/write/error events from several sockets with code like:

```scheme
(zmq-poll (list (poll-item shell-socket) (poll-item iopub-socket)))
```

Ludo'.

PS: GitHub's "merge" button apparently creates gratuitous merge commits so  I'd recommend merging using the `git` command-line tool if possible.  :-)

civodul committed:
Add bindings for 'zmq_poll'.

* src/simple-zmq.scm (zmq_poll): New procedure.
(<poll-item>): New record type.
(poll-item): New procedure.
(ZMQ_POLLIN, ZMQ_POLLOUT, ZMQ_POLLERR, %pollitem-struct): New
variables.
(poll-items->array, array->poll-items, zmq-poll): New procedures.

jerry40 commented:
Hi! Excellent work, thanks! I think there is a high time for writing a test module, Can you give me a point how to do it correctly in a guile way? Also, let me press 'squash and merge' this time since you sent 1 commit. I will figure out the right set of git commands for merging PRs later.

jerry40 commented on src/simple-zmq.scm:
```diff
@@ -197,6 +211,23 @@
+                    #:optional
+                    (events (logior ZMQ_POLLIN ZMQ_POLLOUT ZMQ_POLLERR))
+                    #:key (fd -1))
+  "Return a new \"poll item\" record suitable suitable for 'zmq-poll'.  The
```
A typo "record suitable suitable for"

civodul replied:
oops, indeed!

civodul commented on src/simple-zmq.scm:
```diff
@@ -504,3 +540,72 @@
+                 (cons (poll-item (if (null-pointer? socket)
+                                      #f
+                                      (pointer->socket socket))
+                                  events
```
Another typo: here it should be `revents`, not `events`.  Could you make this change?

civodul committed:
poll: Fix typos.

* src/simple-zmq.scm (poll-item): Fix typo in docstring.
(array->poll-items): Pass REVENTS as the second argument to 'poll-item'
instead of EVENTS.

civodul commented:
With the fixes above it should be alright.  Let me know if you need anything else!

civodul commented:
Regarding tests, I agree it's becoming seriously needed, though it shouldn't hold this PR back IMO.

The way I'd suggest doing it is to:

1. Properly package guile-simple-zmq, for instance using [Hall](https://gitlab.com/a-sassmannshausen/guile-hall), which can generate Autoconf/Automake machinery for you.
2. Add a `tests` directory containing test files that use the `(srfi srfi-64)` test suite library as many projects do.  Here's [an example](https://notabug.org/guile-sqlite3/guile-sqlite3/src/master/tests/basic.scm).
3. Add those tests to `Makefile.am` such that `make check` runs them, as in [this example](https://notabug.org/guile-sqlite3/guile-sqlite3/src/master/Makefile.am).

HTH!

jerry40 commented:
Thanks! This PR is ready to be merged, I'll do this in 3-4 hours (since I'm not ready to use cli to avoid squashing at the moment).

jerry40 commented:
I used cli commands,hopefully commits history is ok now. Thanks!

civodul commented:
Great, thank you!

I personally like to rebase patches on top of the branch rather than
merge them, to keep the history linear as much as possible, but I
understand you might have different preferences.  :-)

jerry40 commented:
At  work we use rebase to master - squash - push force workflow, then one guy just presses 'merge' on bibucket side :) So I'm not very good at merging PRs. I need to read more about this process!
